python/tvm/relay/transform/quantization/quantized_operators/dense.py

```diff
@@ -0,0 +1,183 @@
+from typing import *
+
+import numpy as np
+import tvm
+from tvm import relay
+from tvm.relay.dataflow_pattern import (
+    DFPatternCallback,
+    _DFPatternCallback,
+    is_constant,
+    is_op,
+    wildcard,
+)
+from tvm.relay.op import nn, tensor
+from tvm.relay.transform.quantization import calibration_rewrite_utils
+
+QParams = NamedTuple(
+    "QParams", [("scale_factor", tvm.relay.Expr), ("zero_point", tvm.relay.Expr), ("dtype", str)]
+)
+
+
+class AffineQuantizationVarCreator:
+    """Class which manages references to our qparams and can insert state."""
+
+    def __init__(self):
+        self.ref_count = 0
+        self.qparams = []
+
+    def get_qparams(self, name_hint: str, dtype: str = "int8") -> QParams:
+        scale = relay.Var(f"{name_hint}.scale")
+        zero_point = relay.Var(f"{name_hint}.zero_point")
+        qparam = QParams(scale, zero_point, dtype)
+        self.qparams.append(qparam)
+        self.ref_count += 1
+        return qparam
+
+
+def generate_generic_quantized_dense(
+    data: tvm.relay.Expr,
+    weight: tvm.relay.Expr,
+    data_qparams: QParams,
+    weight_qparams: QParams,
+    internal_accumulation_dtype: str = "float32",
+    simulated_accumulation_dtype: str = "int32",
+    out_units: Optional[int] = None,
+    in_units: Optional[int] = None,
+) -> Tuple[tvm.relay.Expr, QParams]:
+    """TODO"""
+
+    # TODO: figure out whether we need this or we can always have the
```

Comment:
Units is a dense attribute, but it is not always set, so sometimes the caller will be able to pass in the units and sometimes not.

Reply:
I believe that the units in dense is in_units, so the caller won't have access to out_units unless they look at the shape.